website/oauth/models/__init__.py

```diff
@@ -443,3 +443,69 @@ def absolute_api_v2_url(self):
     # used by django and DRF
     def get_absolute_url(self):
         return self.absolute_api_v2_url
+
+class ApiOAuth2PersonalToken(StoredObject):
+    """Information for user-created personal access tokens
+
+    This collection is also used by CAS to create the master list of available tokens.
+    Any changes made to field names in this model must be echoed in the CAS implementation.
+    """
+    _id = fields.StringField(primary=True,
+                             default=lambda: str(ObjectId()))
+
+    # Name of the field being `token_id` is a CAS requirement.
+    # This is the actual value of the token that's used to authenticate
+    token_id = fields.StringField(default=functools.partial(random_string, length=70),
+                               unique=True)
+
+    owner = fields.ForeignField('User',
+                                backref='created',
+                                index=True,
+                                required=True)
+
+    name = fields.StringField(required=True, index=True)
+
+    # This field is a space delimited list of scopes, e.g. "osf.full_read osf.full_write"
+    scopes = fields.StringField(required=True)
+
+    is_active = fields.BooleanField(default=True, index=True)
+
+    def deactivate(self, save=False):
+        """
+        Deactivate an ApiOAuth2PersonalToken
+
+        Does not delete the database record, but hides this instance from API
+        """
+        client = cas.get_client()
+        # Will raise a CasHttpError if deletion fails for any reason other than the token
+        # not yet being created. This will also stop setting of active=False.
+        try:
+            resp = client.revoke_tokens({'token_id': self.token_id})  # noqa
+        except cas.CasHTTPError as e:
+            if e.code == 400:
+                pass  # Token hasn't been used yet, so not created in cas
+            else:
```

Comment:
401 or 403 depending on user authorization being missing or invalid, in which case the error should be raised.

Not that I'm aware of. If it's flagged as `.is_active=False`, CAS won't create it if you attempt to use it after calling this function. If it's already been created, it's able to be deleted in CAS with proper authorization, and without proper authorization it won't be flagged as inactive.